### PR TITLE
Support unsigned types

### PR DESCRIPTION
**Pull request by SvyatoslavScherbina**
To be merged after #1840 and updating to 1.3.
Migrating C interop to unsigned types is to be added later as a separate PR.

### CODE REVIEW

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt [hunk 1]
```diff
@@ -1091,11 +1104,15 @@ private fun objCFunctionType(context: Context, methodBridge: MethodBridge): LLVM
 
 private val ObjCValueType.llvmType: LLVMTypeRef get() = when (this) {
     ObjCValueType.BOOL -> int8Type
+    ObjCValueType.UNICHAR -> int16Type
```
[the remark this answers is not in the document]
Comment (SvyatoslavScherbina) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/llvm/objcexport/ObjCExportCodeGenerator.kt]: Kotlin compiler appears to optimize `when` over enum entries.

File: runtime/src/main/kotlin/kotlin/collections/ArrayUtil.kt [hunk 2]
```diff
@@ -237,6 +237,22 @@ external private fun copyImpl(array: IntArray, fromIndex: Int,
 external private fun copyImpl(array: LongArray, fromIndex: Int,
                               destination: LongArray, toIndex: Int, count: Int)
 
+@SymbolName("Kotlin_ByteArray_copyImpl")
```
Comment (olonho) [inline runtime/src/main/kotlin/kotlin/collections/ArrayUtil.kt]: Maybe add comment that operation is bitwise identical to signed types?

File: runtime/src/main/kotlin/kotlin/collections/ArrayUtil.kt [hunk 3]
```diff
@@ -279,6 +295,30 @@ fun LongArray.copyRangeTo(destination: LongArray, fromIndex: Int, toIndex: Int,
     copyImpl(this, fromIndex, destination, destinationIndex, toIndex - fromIndex)
 }
 
+@SinceKotlin("1.3")
+@ExperimentalUnsignedTypes
```
Comment (olonho) [inline runtime/src/main/kotlin/kotlin/collections/ArrayUtil.kt]: What's the point of this annotation?

File: runtime/src/main/kotlin/kotlin/collections/Arrays.kt [hunk 4]
```diff
@@ -972,6 +1060,54 @@ public actual fun LongArray.contentHashCode(): Int {
...
 
+/**
+ * Returns a hash code based on the contents of this array as if it is [List].
+ */
+@SinceKotlin("1.3")
+@ExperimentalUnsignedTypes
+public fun UByteArray.contentHashCode(): Int {
+    var result = 1
+    for (element in this)
+        result = 31 * result  + element.hashCode()
```
Comment (olonho) [inline runtime/src/main/kotlin/kotlin/collections/Arrays.kt]: Why call `element.hashCode()` and not just `element`?

File: runtime/src/main/kotlin/kotlin/collections/ArrayUtil.kt (hunk 3 above)
Comment (SvyatoslavScherbina) [reply runtime/src/main/kotlin/kotlin/collections/ArrayUtil.kt]: It is recommended to have one on any declaration that uses unsigned types. That's how Kotlin experimental features are organized.

File: runtime/src/main/kotlin/kotlin/collections/Arrays.kt (hunk 4 above)
Comment (SvyatoslavScherbina) [reply runtime/src/main/kotlin/kotlin/collections/Arrays.kt]: 1) Kotlin doesn't have mixed signed-unsigned operations.
2) All other variants of this function call hashCode, including one for ints.

File: runtime/src/main/kotlin/kotlin/collections/ArrayUtil.kt (hunk 3 above)
Comment (olonho) [reply runtime/src/main/kotlin/kotlin/collections/ArrayUtil.kt]: Not sure if it makes any sense for Native.

Comment (SvyatoslavScherbina) [reply runtime/src/main/kotlin/kotlin/collections/ArrayUtil.kt]: I'm sure that it makes sense for Native.

File: runtime/src/main/kotlin/kotlin/collections/ArrayUtil.kt (hunk 2 above)
Comment (SvyatoslavScherbina) [reply runtime/src/main/kotlin/kotlin/collections/ArrayUtil.kt]: Done

**Pull request merged**